stringfuzz/generators/clever.py

```diff
@@ -0,0 +1,356 @@
+    shrunken_depth = random.randint(0, tree_depth - 1)
+
+    # build expression
+    signature   = [sort, sort]
+    random_args = [make_random_tree(variables, arg_sort, shrunken_depth, expr_depth) for arg_sort in signature]
+    random_args = [make_random_expression(variables, BOOL_SORT, expr_depth)] + random_args
+    expression  = _tree_nonterminal(*random_args)
+
+    return expression
+
+def make_clever(max_client_depth, num_client_vars, max_lib_depth, num_lib_vars, num_lib_calls, max_expr_depth, max_str_lit_length, max_int_lit, literal_probability, sorts, sliced, client_name="client", old_lib_name="old_lib", new_lib_name="new_lib"):
+
+    # check args
+    if max_client_depth < 1:
+        raise ValueError('the maximum client depth must be at least 1')
+
+    if max_expr_depth < 1:
+        raise ValueError('the maximum expr depth must be at least 1')
+
+    if num_lib_calls < 1:
+        raise ValueError('the number of library calls must be at least 1')
+
+    if max_lib_depth < 1:
+        raise ValueError('the maximum library depth must be at least 1')
+
+    if len(sorts) < 1 or any(map(lambda x: x not in VARIABLE_SORTS, sorts)):
+        raise ValueError('invalid sorts: {!r}'.format(sorts))
+
+    global _max_str_lit_length
+    global _max_int_lit
+    global _literal_probability
+    global _var_sorts
+    global _expr_nontermianls
+
+    # set global config
+    _max_str_lit_length  = max_str_lit_length
+    _max_int_lit         = max_int_lit
+    _literal_probability = literal_probability
+
+    _var_sorts = []
+    for s in sorts:
+        if s == STRING:
+            _expr_nontermianls.extend(_operators[STRING])
+            _var_sorts.append(STRING_SORT)
+        if s == INT:
+            _expr_nontermianls.extend(_operators[INT])
+            _var_sorts.append(INT_SORT)
+        if s == BOOL:
+            _var_sorts.append(BOOL_SORT)
+
+    # create function definitions
+    lib_sort    = random.choice(_var_sorts)
+    lib_args    = [random.choice(_var_sorts) for s in range(num_lib_vars)]
+    client_sort = random.choice(_var_sorts)
+    client_args = [random.choice(_var_sorts) for s in range(num_client_vars)]
+
+    # create variables
+    arg_var_map = []
+
+    client_vars = {s: [] for s in set(client_args)}
+    client_decls = []
+    for s in client_args:
+        v = smt_new_arg()
+        client_vars[s].append(v)
+        client_decls.append((v,s))
+        arg_var_map.append(v)
+    # create variable declarations
+    client_decls = smt_declare_args(client_decls)
+
+    variables = []
+    decls = []
+    for s in client_args:
+        v = smt_new_var()
+        variables.append(v)
+        decls.append(smt_declare_var(v, sort=s))
+
+    lib_vars = {s: [] for s in set(lib_args)}
+    lib_decls = []
+    for s in lib_args:
+        v = smt_new_arg()
+        lib_vars[s].append(v)
+        lib_decls.append((v,s))
+    # create variable declarations
+    lib_decls = smt_declare_args(lib_decls)
```

Comment:
The above several blocks are big enough to be in their own function; please put them in one.